### PR TITLE
Update TryParse for int to match Utf8Parser

### PR DESCRIPTION
**Pull request by JeremyKuhne**
- Update to handle same inputs Utf8Parser.TryParse handles
- Allow specifying format char
- Update bool try parse

### CODE REVIEW

File: src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs [hunk 1]
```diff
@@ -58,42 +64,88 @@ private unsafe static bool TryParseSlow(ref BufferReader<byte> reader, out bool
...
-                // Fast path had enough space but couldn't find valid data
-                value = default;
-                return false;
-            }
-
-            byte* buffer = stackalloc byte[MaxLength];
-            Span<byte> tempSpan = new Span<byte>(buffer, MaxLength);
-            if (Utf8Parser.TryParse(reader.PeekSlow(tempSpan), out value, out int consumed))
+                // stackalloc is separated as we can't represent that PeekSlow will not capture the span.
+                byte* stackBuffer = stackalloc byte[bufferSize];
```
Comment (KrzysztofCwalina) [inline src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs]: Why do you first create a pointer and then wrap it? you know you can stackalloc span buffers directly?

Comment (JeremyKuhne) [reply src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs]: That is what the comment is about. C# doesn't allow it as you could, in theory, capture the stack span in the reader. We have a request open to allow marking parameters as no capture so we can get around this.

Comment (JeremyKuhne) [reply src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs]: We could also force a copy of the reader by not taking it as ref, but that comes at a non-trivial cost.

File: src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs [hunk 2]
```diff
@@ -58,42 +64,88 @@ private unsafe static bool TryParseSlow(ref BufferReader<byte> reader, out bool
...
+            if (consumed < unread.Length)
             {
+                // The parser found a digit it wouldn't consume so we have all valid data.
                 reader.Advance(consumed);
                 return true;
             }
 
-            return TryParseSlow(ref reader, out value);
+            // If we ate all of our unread there may be more valid digits
+            return TryParseSlow(ref reader, out value, standardFormat);
```
Comment (KrzysztofCwalina) [inline src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs]: This means we will always hit the slow path at the end of the sequence/span, i.e consumed will be equal to unread.Length. Can we avoid that?

Comment (JeremyKuhne) [reply src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs]: Not easy to do as we would need to know if we're in the last span. We could add a check for single segment, but the expectation is that you won't hit this often and we want to keep the fast path as small as possible.

File: src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs [hunk 3]
```diff
@@ -58,42 +64,88 @@ private unsafe static bool TryParseSlow(ref BufferReader<byte> reader, out bool
...
+                Span<byte> buffer = new Span<byte>(stackBuffer, bufferSize);
+                ReadOnlySpan<byte> peekSpan = reader.PeekSlow(buffer);
+
+                if (!Utf8Parser.TryParse(peekSpan, out value, out int consumed, standardFormat))
+                {
+                    // Overflow
+                    return false;
+                }
+
+                if (consumed < peekSpan.Length || peekSpan.Length < buffer.Length)
```
Comment (KrzysztofCwalina) [inline src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs]: what if the int is at the end of the reader's buffer?

File: src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs [hunk 4]
```diff
@@ -58,42 +64,88 @@ private unsafe static bool TryParseSlow(ref BufferReader<byte> reader, out bool
...
+                {
+                    // The parser found a digit it wouldn't consume or we hit the end of the reader
+                    // (peekSpan was smaller than requested)so we have all valid data.
+                    reader.Advance(consumed);
+                    return true;
+                }
+
+                // Throw if doubling would overflow.
+                bufferSize = checked(bufferSize * 2);
+            } while (true);
```
Comment (KrzysztofCwalina) [inline src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs]: should we break out the loop after some threshold?

File: src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs (hunk 3 above)
Comment (JeremyKuhne) [reply src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs]: That is what the `peekSpan.Length < buffer.Length` is checking. If we couldn't fill our requested peek, we're at the end.

File: src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs (hunk 2 above)
Comment (KrzysztofCwalina) [reply src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs]: Maybe the check could be inside Slow method but before we actually allocate the buffer and copy?

File: src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs (hunk 4 above)
Comment (JeremyKuhne) [reply src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs]: We'll throw eventually with an overflow above. I'm not sure we can pick anything more rational than that. We could, in theory, return false- but we're starting to hit crazytown levels here (MBs of leading zeroes). I'm open to suggestions here.

File: src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs (hunk 1 above)
Comment (KrzysztofCwalina) [reply src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs]: Ah, I see.

File: src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs (hunk 2 above)
Comment (JeremyKuhne) [reply src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs]: That we're single segment? We'd still need to parse again- and if we're truly single segment there will be no copy- peek will just slice.

File: src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs (hunk 3 above)
Comment (KrzysztofCwalina) [reply src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs]: I was wondering what happens if we request 15 bytes, get 15 bytes, and then parsed exactly 15 bytes? I gather we just loop and then the code path you are referring to will be hit. But this, together with my comment in the outer TryParse, means that if I parse exactly 15 bytes int, I will not only hit the slow code path, but also loop here. If it's not avoidable, then maybe it's ok, but it does seem wasteful at the first glance.

File: src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs (hunk 4 above)
Comment (KrzysztofCwalina) [reply src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs]: What does int.Parse do? Does it accept 1 billion leading zeros?

File: src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs (hunk 2 above)
Comment (KrzysztofCwalina) [reply src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs]: Ah, I see. I forgot that peek with simply return the last span (if it's single segment)

File: src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs (hunk 4 above)
Comment (JeremyKuhne) [reply src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs]: Well, up to almost int.MaxValue in theory. I considered doing our own scanning to get past these cases, but it is complicated and would fall over if the Utf8Parser changes behavior or adds another format type. There are a number of corner cases in that code.

I think it is reasonable to expect users to do their own skipping for crazy cases. We've got all the helpers for that.

File: src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs (hunk 3 above)
Comment (JeremyKuhne) [reply src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs]: You'd have to have exactly 15 bytes left in the reader's remaining span **and** have them all be valid to hit this case. Normal cases will be < 15. Together I think that case should be really rare.

You can also come in with more than 15 bytes if you're matching the end of the span. Same sort of rareness. We could change behavior here to check for these cases to avoid extra work, but I would expect them to impact the perf of common case- which is 14 bytes crossing segments.

File: src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs [hunk 5]
```diff
@@ -58,42 +64,88 @@ private unsafe static bool TryParseSlow(ref BufferReader<byte> reader, out bool
...
             ReadOnlySpan<byte> unread = reader.UnreadSpan;
-            if (Utf8Parser.TryParse(unread, out value, out int consumed) && consumed < unread.Length)
+            if (!Utf8Parser.TryParse(unread, out value, out int consumed, standardFormat))
+            {
+                return false;
+            }
+
+            if (consumed < unread.Length)
             {
+                // The parser found a digit it wouldn't consume so we have all valid data.
```
Comment (ahsonkhan) [inline src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs]: How does this comment explain what's happening here? How do we know that the parser found a digit, specifically? Isn't it just that there is more data left in this segment?

File: src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs [hunk 6]
```diff
@@ -58,42 +64,88 @@ private unsafe static bool TryParseSlow(ref BufferReader<byte> reader, out bool
...
             {
-                // Fast path had enough space but couldn't find valid data
-                value = default;
-                return false;
-            }
-
-            byte* buffer = stackalloc byte[MaxLength];
-            Span<byte> tempSpan = new Span<byte>(buffer, MaxLength);
-            if (Utf8Parser.TryParse(reader.PeekSlow(tempSpan), out value, out int consumed))
+                // stackalloc is separated as we can't represent that PeekSlow will not capture the span.
```
Comment (ahsonkhan) [inline src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs]: represent -> guarantee?

Comment (JeremyKuhne) [reply src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs]: represent- we can technically guarantee, we just have no way to represent that to C# (yet)

File: src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs [hunk 7]
```diff
@@ -58,42 +64,88 @@ private unsafe static bool TryParseSlow(ref BufferReader<byte> reader, out bool
...
-
-            ReadOnlySpan<byte> unread = reader.UnreadSpan;
+            // Note that ints can have leading zeros ("-00000000000000127") which makes predicting the max
+            // length difficult. Typically the int will be less than 15 characters ("-2,147,483,648") so we'll
+            // work in those increments.
+            int bufferSize = 15;
 
-            if (unread.Length > MaxLength)
+            // Try using the stack for scratch space.
+            do
```
Comment (ahsonkhan) [inline src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs]: Rather than continuously doubling (and potentially throwing if there are 2.1 billion leading zeroes, which should be valid), why not skip all leading zeroes ahead of time? In the common case, it will be a single check but the implementation would be clearer.

Comment (JeremyKuhne) [reply src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs]: We'd have to deal with '+', '-', '.' and other formats that might come in the future.  Doesn't seem wise to couple that closely and add that much complexity for something that arguably should be handled by the consumer of this API.

That said, I think might even be more aggressive here. Lets say someone throws an empty 1GB buffer at us- should we really cruise the entire thing and return 0? Is that really what a user would want?

**Comment by JeremyKuhne** [comment]
Marked as no merge temporarily. Found an issue where we fall over reading segments that end with a single `+`, `-`, or `.`.

File: tests/System.Buffers.Experimental.Tests/BufferFactory.cs [hunk 8]
```diff
@@ -53,6 +53,14 @@ public static ReadOnlySequence<byte> Create(params byte[][] buffers)
             return Create(list.ToArray());
         }
 
+        public static ReadOnlySequence<byte> CreateUtf8(params string[] buffers)
+        {
+            if (buffers.Length == 1) return new ReadOnlySequence<byte>(Encoding.UTF8.GetBytes(buffers[0]));
+            var list = new List<Memory<byte>>();
+            foreach (var b in buffers) list.Add(Encoding.UTF8.GetBytes(b));
```
Comment (ahsonkhan) [inline tests/System.Buffers.Experimental.Tests/BufferFactory.cs]: nit: avoid using var here

also not: avoid single letter variable names

File: tests/System.Buffers.Primitives.Tests/BufferUtilities.cs [hunk 9]
```diff
@@ -114,5 +115,20 @@ public static ReadOnlySequence<byte> CreateUtf8Buffer(params string[] inputs)
             }
             return CreateBuffer<T>(buffers);
         }
+
+        public static void FillIntegerUtf8Array(byte[] array, int minValue, int maxValue, int seed = 42)
+        {
+            Random r = new Random(42);
```
Comment (ahsonkhan) [inline tests/System.Buffers.Primitives.Tests/BufferUtilities.cs]: Use the seed that was passed in.

File: tests/System.Buffers.ReaderWriter.Tests/Reader_ParseFloatingPoint.cs [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+using System.Buffers.Reader;
+using System.Globalization;
+using System.Text;
+using Xunit;
+
+namespace System.Buffers.Tests
+{
+    public class Reader_ParseFloatingPoint
+    {
+
```
Comment (ahsonkhan) [inline tests/System.Buffers.ReaderWriter.Tests/Reader_ParseFloatingPoint.cs]: nit: remove extra white space

File: src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs [hunk 11]
```diff
@@ -3,22 +3,57 @@
...
+        private static ParseDelegate<uint> s_uintParse;
         private static ParseDelegate<long> s_longParse;
+        private static ParseDelegate<ulong> s_ulongParse;
+        private static ParseDelegate<float> s_floatParse;
+        private static ParseDelegate<double> s_doubleParse;
+        private static ParseDelegate<decimal> s_decimalParse;
+
+        private static ParseDelegate<byte> ByteParser
+        {
+            [MethodImpl(MethodImplOptions.NoInlining)]
```
Comment (ahsonkhan) [inline src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs]: Why is this necessary?

File: src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs [hunk 12]
```diff
@@ -56,9 +159,9 @@ private static ParseDelegate<long> LongParser
         /// <see cref="Utf8Parser.TryParse(ReadOnlySpan{byte}, out bool, out int, char)"/> for supported formats.
         /// </remarks>
         /// <exception cref="FormatException">Thrown if the given format isn't supported.</exception>
-        /// <returns>True if successfully parsed.</returns>
+        /// <returns>Number of consumed bytes if successful, otherwise 0.</returns>
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public static unsafe bool TryParse(ref this BufferReader<byte> reader, out bool value, char standardFormat = '\0')
+        public static unsafe int TryParse(ref this BufferReader<byte> reader, out bool value, char standardFormat = '\0')
```
Comment (ahsonkhan) [inline src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs]: This method doesn't need to be unsafe.

Comment (ahsonkhan) [reply src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs]: Also, do we have precedence elsewhere that a TryX method doesn't return bool? If not, maybe we should rename it.

File: src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs [hunk 13]
```diff
@@ -68,91 +171,152 @@ public static unsafe bool TryParse(ref this BufferReader<byte> reader, out bool
             if (Utf8Parser.TryParse(unread, out value, out int consumed, standardFormat))
             {
                 reader.Advance(consumed);
-                return true;
+                return consumed;
+            }
+            else if (unread.Length >= MaxBoolLength)
```
Comment (ahsonkhan) [inline src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs]: I am not sure if this is correct. What if `unread = {tru1}`? TryParse returns false, `unread.Length is < MaxBoolLength`, and then we fall to TryParseSlow which isn't what we want to do here.

File: src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs [hunk 14]
```diff
@@ -3,138 +3,700 @@
...
+        /// </summary>
+        /// <param name="value">The parsed <see cref="byte"/> value or default if failed.</param>
+        /// <param name="standardFormat">Expected format.</param>
+        /// <remarks>
+        /// <see cref="Utf8Parser.TryParse(ReadOnlySpan{byte}, out byte, out int, char)"/> for supported formats.
+        /// </remarks>
+        /// <exception cref="FormatException">Thrown if the given format isn't supported.</exception>
+        /// <returns>Number of consumed bytes if successful, otherwise 0.</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static unsafe int TryParse(
```
Comment (ahsonkhan) [inline src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs]: same here, doesn't need to be unsafe

File: src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs [hunk 15]
```diff
@@ -3,138 +3,700 @@
...
+            int bufferSize,
+            ParseDelegate<T> parseDelegate) where T: unmanaged
+        {
+            // Note that in calling this with the expected max size of a normal instance of the given
+            // type may do more work than needed for the degenerate case of zero prefaced numbers at
+            // the end of the span (e.g. 00000000000000000000000000123). Doing the extra check slows
+            // down the perf of "normal" numbers so we just let the "bad" numbers eat up more cycles.
+
+            // stackalloc is separated as we can't represent in C# that Peek will not capture the span.
+            byte* stackBytes = stackalloc byte[MaxParseBytes];
```
Comment (ahsonkhan) [inline src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs]: Does it make sense to separate out the common path (where we stackalloc exactly bufferSize rather than MaxParseBytes)? If we fail on our first attempt, we can then fall back to the doubling loop below.

File: src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs [hunk 16]
```diff
@@ -3,138 +3,700 @@
...
+        /// <exception cref="FormatException">Thrown if the given format isn't supported.</exception>
+        /// <returns>Number of consumed bytes if successful, otherwise 0.</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static unsafe int TryParse(
+            ref this BufferReader<byte> reader,
+            out int value,
+            char standardFormat = '\0')
         {
-            const int MaxLength = 15;
+            //return TryParseFast(ref reader, out value, standardFormat, 15, IntParser);
```
Comment (ahsonkhan) [inline src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs]: remove commented out code

File: src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs (hunk 13 above)
Comment (JeremyKuhne) [reply src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs]: It is what we want to do, if we want this code path to be as fast as possible. Every check shows up in the perf runs. All of these methods make some technically unnecessary trips down the slow path.

File: src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs [hunk 17]
```diff
@@ -68,91 +171,152 @@ public static unsafe bool TryParse(ref this BufferReader<byte> reader, out bool
...
+            }
 
-                if (!parseDelegate(peekSpan, out value, out int consumed, standardFormat))
-                {
-                    // Overflow
-                    return false;
-                }
+            // If we ate all of our unread there may be more valid digits.
+            // The typical max size for a short will be less than 4 bytes ("255").
+            return TryParseSlow(ref reader, out value, standardFormat, 4, ByteParser);
```
[the remark this answers is not in the document]
Comment (JeremyKuhne) [reply src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs]: Because you need a non-valid character after the valid data.

File: src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs (hunk 11 above)
Comment (JeremyKuhne) [reply src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs]: Just being explicit- don't ever want to generate the delegate until we absolutely need it.

File: src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs (hunk 15 above)
Comment (JeremyKuhne) [reply src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs]: No, because the logic would be really complicated to avoid stack pressure. Every `stackalloc` adds more to the stack, even in a loop.

File: src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs (hunk 16 above)
Comment (JeremyKuhne) [reply src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs]: Whoops- remnant of discovering there is no way to do generics here without significant overhead. :(

File: tests/System.Buffers.Experimental.Tests/BufferFactory.cs (hunk 8 above)
Comment (JeremyKuhne) [reply tests/System.Buffers.Experimental.Tests/BufferFactory.cs]: Yeah, copying existing code around.

Comment (JeremyKuhne) [reply tests/System.Buffers.Experimental.Tests/BufferFactory.cs]: Var is ok in this context btw

File: tests/System.Buffers.Primitives.Tests/BufferUtilities.cs (hunk 9 above)
Comment (JeremyKuhne) [reply tests/System.Buffers.Primitives.Tests/BufferUtilities.cs]: Derp

File: src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs [hunk 18]
```diff
@@ -3,138 +3,700 @@
...
+            [MethodImpl(MethodImplOptions.NoInlining)]
+            get
+            {
+                if (s_intParse == null)
+                    s_intParse = Utf8Parser.TryParse;
+                return s_intParse;
+            }
+        }
+
+        private static ParseDelegate<uint> UintParser
```
Comment (KrzysztofCwalina) [inline src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs]: nit:lower case "i"

File: src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs [hunk 19]
```diff
@@ -3,138 +3,700 @@
...
+            [MethodImpl(MethodImplOptions.NoInlining)]
+            get
+            {
+                if (s_byteParse == null)
+                    s_byteParse = Utf8Parser.TryParse;
+                return s_byteParse;
+            }
+        }
+
+        private static ParseDelegate<sbyte> SbyteParser
```
Comment (KrzysztofCwalina) [inline src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs]: nit: I would call it "SByte" to match the type name

File: src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs [hunk 20]
```diff
@@ -3,138 +3,700 @@
...
+        /// </summary>
+        /// <param name="value">The parsed <see cref="float"/> value or default if failed.</param>
+        /// <param name="standardFormat">Expected format.</param>
+        /// <remarks>
+        /// <see cref="Utf8Parser.TryParse(ReadOnlySpan{byte}, out float, out int, char)"/> for supported formats.
+        /// </remarks>
+        /// <exception cref="FormatException">Thrown if the given format isn't supported.</exception>
+        /// <returns>Number of consumed bytes if successful, otherwise 0.</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static unsafe int TryParse(
```
Comment (KrzysztofCwalina) [inline src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs]: Why did we return number of consumed bytes as opposed to having an out parameter? I assume it's for perf, but it's not consistent with the general pattern that we already shipped. Is it really worth departing from the pattern?

Comment (JeremyKuhne) [reply src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs]: Yeah, but it wasn't a huge difference. I'm going to change it.

File: src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs (hunk 19 above)
Comment (JeremyKuhne) [reply src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs]: I was matching the c# keywords. I can change these to Single etc.
